spmd/tensor/ops/tensor_ops.py

```diff
@@ -278,3 +279,176 @@ def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
                 )
             ],
         )
+
+
+@register_prop_rule("aten.index_select.default")
+def prop_index_select(op_schema: OpSchema) -> OutputSharding:
+    values_spec, dim, indices_spec = op_schema.args_schema
+
+    assert isinstance(values_spec, DTensorSpec)
+    assert isinstance(dim, int)
+    assert isinstance(indices_spec, DTensorSpec)
+
+    all_indices_spec: List[Optional[DTensorSpec]] = [
+        indices_spec if dim == i else None for i in range(values_spec.ndim)
+    ]
+
+    result = prop_index(
+        OpSchema(
+            args_schema=(values_spec, all_indices_spec),
+            kwargs_schema=op_schema.kwargs_schema,
+        )
+    )
+    if result.schema_suggestions:
+        result.schema_suggestions = [
+            OpSchema(
+                args_schema=(s.args_schema[0], dim, s.args_schema[1][dim]),
+                kwargs_schema=op_schema.kwargs_schema,
+            )
+            for s in result.schema_suggestions
+        ]
+    return result
+
+
+@register_prop_rule("aten.index.Tensor")
+def prop_index(op_schema: OpSchema) -> OutputSharding:
+    """
+    Expect replicated on the first input; _mostly_ pointwise on the second input.
+    TODO: exception: when the dtype of second input is "bool", then a torch.nonzero needs to be triggered first.
+    """
+    values_spec, multi_indices_spec = op_schema.args_schema
+    assert isinstance(values_spec, DTensorSpec)
+    multi_indices_spec = cast(
+        Union[DTensorSpec, List[Optional[DTensorSpec]]], multi_indices_spec
+    )
+
+    if isinstance(multi_indices_spec, list):
+        valid_indices_spec: List[Tuple[int, DTensorSpec]] = [
+            (i, a) for i, a in enumerate(multi_indices_spec) if a is not None
+        ]
+    else:
+        valid_indices_spec = [(0, multi_indices_spec)]
```

Comment:
what's this "0" for, `multi_indices_spec` is non-zero, so we need to prepend 0 for indexing in this case?

Reply:
Oh I see, it's the default dim to index into?

Reply:
Let me take this back. I don't think this is actually a valid input. I removed this and it simplifies things.